openfermioncirq/gates/four_qubit_gates.py

```diff
@@ -0,0 +1,165 @@
+            degs: The exponent angle, in degrees.
+            duration: The exponent as a duration of time.
+        """
+
+        if len([1 for e in [half_turns, rads, degs, duration]
+                if e is not None]) > 1:
+            raise ValueError('Redundant exponent specification. '
+                             'Use ONE of half_turns, rads, degs, or duration.')
+
+        if duration is not None:
+            exponent = 2 * duration / numpy.pi
+        else:
+            exponent = cirq.value.chosen_angle_to_half_turns(
+                half_turns=half_turns,
+                rads=rads,
+                degs=degs)
+
+        super().__init__(exponent=exponent)
+
+    @property
+    def half_turns(self) -> Union[cirq.Symbol, float]:
+        return self._exponent
+
+    def _eigen_components(self):
+        minus_one_component = numpy.zeros((16, 16))
+        minus_one_component[3, 3] = minus_one_component[12, 12] = 0.5
+        minus_one_component[3, 12] = minus_one_component[12, 3] = 0.5
+
+        plus_one_component = numpy.zeros((16, 16))
+        plus_one_component[3, 3] = plus_one_component[12, 12] = 0.5
+        plus_one_component[3, 12] = plus_one_component[12, 3] = -0.5
+
+        return [
+            (-1, minus_one_component),
+            (1, plus_one_component),
+            (0, numpy.diag([1, 1, 1, 0, 1, 1, 1, 1,
+                            1, 1, 1, 1, 0, 1, 1, 1]))
+        ]
+
+    def _canonical_exponent_period(self) -> Optional[float]:
+        return 2
+
+    def _with_exponent(self,
+                       exponent: Union[cirq.Symbol, float]
+                       ) -> 'LocalPQRSGate':
+        return LocalPQRSGate(half_turns=exponent)
+
+    def default_decompose(self, qubits):
+        p, q, r, s = qubits
+
+        yield cirq.CNOT(r, s)
+        yield cirq.CNOT(q, p)
+        yield cirq.CNOT(q, r)
+
+        yield cirq.X(q) ** self.half_turns
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+
+        yield cirq.CNOT(s, r)
+        yield cirq.CNOT(r, q)
+        yield cirq.CNOT(s, r)
+
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+        yield cirq.X(p)
+        yield cirq.CNOT(p, q)
+        yield cirq.X(p)
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+
+        yield cirq.CNOT(s, r)
+        yield cirq.CNOT(r, q)
+        yield cirq.CNOT(s, r)
+
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+        yield cirq.X(q) ** -self.half_turns
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+
+        yield cirq.CNOT(s, r)
+        yield cirq.CNOT(r, q)
+        yield cirq.CNOT(s, r)
+
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+        yield cirq.X(p)
```

Comment:
You can do one X on the target instead of two on the controls to invert the CNOT.

Reply:
done